### ISSUE TITLE
What is the best way to "crawl" a repository?

### ISSUE THREAD

**Issue by mjordan** [mjordan/riprap#6]
The "fetchresourcelist" plugin gets a list of resource URLs to check. This is just a demo. In production, we'll need a way to check every object in the repository. What's the best way to do this?

* Start at the smallest node ID in the Drupal database and page through it?
* Perform a similar SPARQL query against the triplestore?
* Perform a Solr query?

Riprap's plugin architecture will allow us to do all of these. What are people's preferences?

**Comment by whikloj** [comment]
Probably a good idea to move along a pre-configured path. So for Fedora it could be following `ldp:contains` from down through the repo.

**Comment by ajs6f** [comment]
That would work for any LDP impl, although some may not guarantee that the whole repo graph is connected (Trellis does not). It is also excellently parallelize-able.

**Comment by ajs6f** [comment]
Yep, what lets you parallelize fully is the crawling bit-- the actual mechanism by which you get the next resource is totally an impl detail. I"m not saying that crawling is the _only_ way to keep fully parallel, just that it's pretty convenient, easy to understand, and there are a variety of off-the-shelf crawlers and web spiders out there for reuse.

**Comment by mjordan** [comment]
As I was saying during the last CLAW call, I had imagined paging through the repository, fixity-checking a page (say 1000) per scheduled job. That's the way that 7.x's Checksum Checker works now, and allows the repo administrator to spread the load of the checks over time. What options for paging through the LDP container are there, e.g. using offset/limit pages? Also, in [this comment](https://github.com/Islandora-CLAW/CLAW/issues/396#issuecomment-416258516) @ajs6f mentions that the community implementation of Fedora currently suffers from the many-members problem. If we page through the container, do me mitigate that issue?

@whikloj yes, I built in the plugin architecture admins can enable the 'fetchresourcelist' plugin that best suites their repo and resources. So we're not tied to one way of doing this.

**Comment by ajs6f** [comment]
@mjordan Paging. Oh, goodness, gracious, _paging_. It's hard enough that the LDP Working Group could not get working impls to confirm a recommendation as part of their effort and instead offered a [note](https://www.w3.org/2012/ldp/hg/ldp-paging.html). I honestly do not know how well implemented it is today. @acoburn, do we have that for Trellis?

Besides paging, though, for this particular application an HTTP response that streams well might be good enough. That's a fairly low bar to set for the backend.

**Comment by mjordan** [comment]
Yes, agreed, paging is a headache. The current plugin is just a proof of concept (it reads resource URLs from a file, and those URLs exist in the bundled mock Fedora-spec-compliant repository) but I'd advocate for making the first production plugin as configure-and-forget for as many early CLAW implementers. I'm not sure what that plugin would look like but please keep the ideas coming....

**Comment by mjordan** [comment]
@acoburn thanks for the detailed overview of how to simulate paging. Very useful.
